File: pipeline/aeb/EventDetector.m
classdef EventDetector
    properties
        pathToMat                 % original working directory
        pathToMatChunks           % destination folder for PostProcessing
        
        % Parameters for event extraction
        PRE_TIME = seconds(4)     % time before event (duration)
        POST_TIME = seconds(3)    % time after event (duration)
    end
    
    methods
        %% Constructor
        function obj = EventDetector(inputHandler)
            % Constructor: accepts an InputHandler instance
            % Parameters PRE_TIME and POST_TIME are set to defaults
            
            if nargin < 1
                error('InputHandler instance required for initialization.');
            end
            
            % Validate input is an InputHandler instance
            if ~isa(inputHandler, 'InputHandler')
                error('Input argument must be an instance of InputHandler.');
            end
            
            % Set paths using InputHandler's pathToRawData
            obj.pathToMat       = inputHandler.pathToRawData;
            obj.pathToMatChunks = fullfile(obj.pathToMat, 'PostProcessing');
            
            % Create PostProcessing directory if it doesn't exist
            if ~exist(obj.pathToMatChunks, 'dir')
                mkdir(obj.pathToMatChunks);
            end
            
            % Set default parameters (already set as property defaults, but explicit for clarity)
            obj.PRE_TIME  = seconds(4);
            obj.POST_TIME = seconds(3);
        end % constructor
        
        %% Main function to process all .mat files
        % Expects each .mat file to contain a variable 'signalMat'  
        % which is a timetable with required signals
        % Calls detectEvents and extractAEBEvents   
        % Saves extracted chunks to PostProcessing folder
        % Each chunk is saved as <original_filename>_<event_index>.mat
        function processAllFiles(obj)
            
            cd(obj.pathToMat);
            files = dir('*.mat');
            N = length(files);

            for i = 1:N
                filePath = fullfile(files(i).folder, files(i).name);
                [~, name, ~] = fileparts(filePath);
                % Load the .mat file
                matFile = load(filePath, 'signalMat');
               
                signalMat = matFile.signalMat;

                % Force every field into a column vector
                fields = fieldnames(signalMat);
                for f = 1:numel(fields)
                    signalMat.(fields{f}) = signalMat.(fields{f})(:);
                end


                % Process this file
                obj.processSingleFileInternally(signalMat, name);
            end
            cd(obj.pathToMat);
        end
        
        %% Process a single file
        function processSingleFileInternally(obj, signalMat, name)
            % Run event detection
            [locateStartTime, locateEndTime] = obj.detectEvents(signalMat);

            % Extract AEB chunks
            obj.extractAEBEvents(signalMat, locateStartTime, locateEndTime, name);
        end
        
        %% Detect start and end times of AEB events
        function [locateStartTime, locateEndTime] = detectEvents(~, signalMat)
            modeChange      = diff(signalMat.aebTargetDecel);
            locateStart     = find(diff(modeChange < -30));
            locateStartReq  = signalMat.aebTargetDecel(locateStart + 1, 1) < -5.9;
            locateStartTime = signalMat.time(locateStart, 1) .* locateStartReq;
            nonZeroTimes    = nonzeros(locateStartTime);
            if isempty(nonZeroTimes)
                locateStartTime = duration.empty;
            else
                locateStartTime = seconds(nonZeroTimes);
            end
            
            locateEnd       = find(modeChange > 20);
            locateEndStat   = signalMat.aebTargetDecel(locateEnd + 1, 1) > 20;
            locateEndTime   = signalMat.time(locateEnd, 1) .* locateEndStat;
            nonZeroEndTimes = nonzeros(locateEndTime);
            if isempty(nonZeroEndTimes)
                if ~isempty(locateStartTime)
                    locateEndTime = locateStartTime + seconds(7);
                else
                    locateEndTime = duration.empty;
                end
            else
                locateEndTime = seconds(nonZeroEndTimes);
            end
        end % detectEvents
        
        %% Extract AEB events (-PRE_TIME to +POST_TIME) and save chunks
        function extractAEBEvents(obj, signalMat, locateStartTime, locateEndTime, name)
            loopRun = length(locateStartTime);
            fields  = fieldnames(signalMat);
            
            % Validate field sizes
            expectedRows = size(signalMat.time, 1);
            for f = 1:length(fields)
                if size(signalMat.(fields{f}), 1) ~= expectedRows
                    error('Field %s has %d rows, expected %d', fields{f}, size(signalMat.(fields{f}), 1), expectedRows);
                end
            end
            
            for j = 1:loopRun
                % Convert startT to numeric seconds
                startT          = locateStartTime(j) - obj.PRE_TIME;
                startT_seconds  = seconds(startT);
                [~, start]      = utils.findDuration(signalMat.time, startT_seconds);

                % Compute startNext
                if j + 1 <= loopRun
                    startNextT          = locateStartTime(j + 1) - obj.PRE_TIME;
                    startNextT_seconds  = seconds(startNextT);
                    [~, startNext]      = utils.findDuration(signalMat.time, startNextT_seconds);
                else
                    startNext = length(signalMat.time);
                end

                % Initialize stop
                stop = 0;
                n = 0;

                % Try to find a valid end time
                if j + n <= length(locateEndTime)
                    while stop < start && j + n <= length(locateEndTime)
                        stopT = locateEndTime(j + n) + obj.POST_TIME;
                        stopT_seconds = seconds(stopT);
                        [~, stop] = utils.findDuration(signalMat.time, stopT_seconds);
                        if stop < start || stop > startNext
                            n = n + 1;
                        else
                            break;
                        end
                    end
                end

                % Fallback if no valid end time found
                if stop < start || j + n > length(locateEndTime)
                    stopT         = locateStartTime(j) + obj.POST_TIME;
                    stopT_seconds = seconds(stopT);
                    [~, stop]     = utils.findDuration(signalMat.time, stopT_seconds);
                end

                % Ensure stop doesn't exceed signal length
                stop = min(stop, length(signalMat.time));

                if stop > start
                    % Create signalMatChunk as a struct
                    signalMatChunk = struct();
                    for f = 1:length(fields)
                        try
                            signalMatChunk.(fields{f}) = signalMat.(fields{f})(start:stop);
                        catch e
                            error('Error indexing field %s: %s', fields{f}, e.message);
                        end
                    end

                    % Save directly into pathToMatChunks, overwrite if exists
                    extractFileName = fullfile(obj.pathToMatChunks, sprintf('%s_%d.mat', name, j));
                    try
                        save(extractFileName, 'signalMatChunk');
                    catch e
                        error('Error saving %s: %s', extractFileName, e.message);
                    end
                end

            end % for j = 1:loopRun
        end % extractAEBEvents
    end % methods
end % classdef

File: pipeline/aeb/event_detector.py
import os
import gc
import warnings
import numpy as np
import pandas as pd
from utils.signal_mdf import SignalMDF


class AebEventDetector:
    """
    AebEventDetector
    -------------
    Detects AEB events in MF4 files and extracts event chunks.

    Each MF4 file must contain a signal 'aebTargetDecel'.
    """

    # --- Fallback defaults (used only if not defined in config) ---
    PRE_TIME_AEB          = 6.0    # seconds before event
    POST_TIME_AEB         = 3.0    # seconds after event
    START_DECEL_DELTA     = -30.0  # threshold for large negative change (AEB start)
    END_DECEL_DELTA       = 29.0   # threshold for positive change (AEB end)
    PB_TGT_DECEL          = -6.0   # threshold for AEB active state

    def __init__(self, input_handler, config=None):
        if input_handler is None or not hasattr(input_handler, "path_to_raw_data"):
            raise TypeError("AebEventDetector requires an InputHandler instance.")

        self.path_to_mdf = input_handler.path_to_raw_data
        self.path_to_aeb_chunks = os.path.join(self.path_to_mdf, "aeb_chunks")

        if not os.path.exists(self.path_to_aeb_chunks):
            os.makedirs(self.path_to_aeb_chunks)

        # --- Load parameters from config if available ---
        if config is not None and hasattr(config, "params"):
            try:
                params = config.params or {}

                self.pre_time          = float(params.get("PRE_TIME_AEB", self.PRE_TIME_AEB))
                self.post_time         = float(params.get("POST_TIME_AEB", self.POST_TIME_AEB))
                self.start_decel_delta = float(params.get("START_DECEL_DELTA", self.START_DECEL_DELTA))
                self.end_decel_delta   = float(params.get("END_DECEL_DELTA", self.END_DECEL_DELTA))
                self.pb_tgt_decel      = float(params.get("PB_TGT_DECEL", self.PB_TGT_DECEL))

                print(
                    f"⏱ Using PRE={self.pre_time}s, POST={self.post_time}s, "
                    f"Δstart={self.start_decel_delta}, Δend={self.end_decel_delta} from config."
                )

            except Exception as e:
                warnings.warn(f"⚠️ Failed to load timing/delta params from config: {e}")
                self._apply_defaults()
        else:
            self._apply_defaults()

    def _apply_defaults(self):
        """Fallback to class defaults."""
        self.pre_time           = self.PRE_TIME_AEB
        self.post_time          = self.POST_TIME_AEB
        self.start_decel_delta  = self.START_DECEL_DELTA
        self.end_decel_delta    = self.END_DECEL_DELTA
        self.pb_tgt_decel       = self.PB_TGT_DECEL
        print(
            f"⚙️ Using default parameters: PRE={self.pre_time}s, POST={self.post_time}s, "
            f"Δstart={self.start_decel_delta}, Δend={self.end_decel_delta}"
        )

    # -------------------- Public API -------------------- #

    def process_all_files(self, pre_time: float = None, post_time: float = None):
        if pre_time is not None:
            self.pre_time  = float(pre_time)
        if post_time is not None:
            self.post_time = float(post_time)

        # Only process *_extracted.mf4 files
        mdf_files = [f for f in os.listdir(self.path_to_mdf) if f.endswith("_extracted.mf4")]
        print(f"\n📂 Found {len(mdf_files)} extracted MF4 files in {self.path_to_mdf}\n")

        if not mdf_files:
            print("⚠️ No extracted MF4 files found. Did you run InputHandler.process_mf4_files first?")
            return

        for fname in mdf_files:
            file_path = os.path.join(self.path_to_mdf, fname)
            name, _   = os.path.splitext(fname)

            try:
                print(f"🔍 Reading MF4 file: {fname}")
                # ✅ Use SignalMDF instead of raw MDF
                mdf = SignalMDF(file_path)

                # ✅ Access via dot-notation
                if mdf.aebTargetDecel.size == 0:
                    print(f"⚠️ File {fname} missing required signal 'aebTargetDecel' → skipped")
                    continue

                df = pd.DataFrame({
                    "time": mdf.time,
                    "aebTargetDecel": mdf.aebTargetDecel
                })

                start_times, end_times = self.detect_aeb_events(df)
                print(f"   ➝ Detected {len(start_times)} events")

                self.extract_aeb_events(mdf, start_times, end_times, name)

                del mdf, df
                gc.collect()

            except Exception as e:
                print(f"⚠️ Error processing {fname}: {e}")

    # -------------------- Event Detection -------------------- #

    def detect_aeb_events(self, df: pd.DataFrame):
        """Detect start and end times of AEB events."""
        time             = df["time"].values
        aeb_target_decel = df["aebTargetDecel"].values
        decel_delta      = np.diff(aeb_target_decel)

        # Start events
        locate_start     = np.where(np.diff(decel_delta < self.start_decel_delta))[0]
        start_mask       = aeb_target_decel[locate_start + 1] <= self.pb_tgt_decel
        start_times      = time[locate_start][start_mask]

        # End events
        locate_end = np.where(decel_delta > self.end_decel_delta)[0]
        end_times  = time[locate_end]

        if len(end_times) == 0 and len(start_times) > 0:
            buffer_time = max(1.0, self.post_time + 4)
            end_times   = start_times + buffer_time

        return start_times, end_times

    # -------------------- Event Extraction -------------------- #

    def extract_aeb_events(self, mdf: SignalMDF, start_times, end_times, name: str):
        """Extract AEB event chunks and save into mdf_chunks folder (MF4 only)."""

        if len(start_times) == 0:
            print(f"⚠️ No events detected for {name}")
            return

        sig = mdf.aebTargetDecel
        if sig.size == 0:
            print(f"⚠️ Missing 'aebTargetDecel' samples → cannot extract chunks.")
            return

        t_min, t_max = float(mdf.time[0]), float(mdf.time[-1])
        print(f"   🕒 File time range: {t_min:.3f}s → {t_max:.3f}s")

        for j, start_time in enumerate(start_times):
            start_sec = max(start_time - self.pre_time, t_min)
            stop_sec  = (
                min(end_times[j] + self.post_time, t_max)
                if j < len(end_times)
                else min(start_time + self.post_time, t_max)
            )

            if stop_sec <= start_sec:
                print(f"⚠️ Skipping invalid window {start_sec:.3f}s → {stop_sec:.3f}s")
                continue

            print(f"   [DEBUG] cut window: {start_sec:.3f}s → {stop_sec:.3f}s")
            try:
                mdf_chunk = mdf.cut(start=start_sec, stop=stop_sec)
                if not mdf_chunk or len(mdf_chunk.channels_db) == 0:
                    print(f"⚠️ Empty chunk ({start_sec:.2f}s–{stop_sec:.2f}s) → no overlap.")
                    continue

                mf4_name = f"{name}_aeb_{j+1:02d}.mf4"
                mf4_path = os.path.join(self.path_to_aeb_chunks, mf4_name)
                mdf_chunk.save(mf4_path, overwrite=True)
                print(f"   ✅ Saved event {j+1}: {start_sec:.2f}s → {stop_sec:.2f}s → {mf4_name}")

                del mdf_chunk
                gc.collect()

            except Exception as e:
                print(f"⚠️ Error saving event {j+1} of {name}: {e}")

